### ISSUE TITLE
The `no_echo` option does not actually appear to prevent echoes

### ISSUE THREAD

**Issue by Jess3Jane** [nats-io/nats.rs#147]
Make sure that these boxes are checked before submitting your issue -- thank you!

 - [x] Included below version and environment information
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Steps or code to reproduce the issue: 

```rust
use async_nats::Options;

#[tokio::main]
async fn main() {
	let connection = Options::new()
		.no_echo()
		.connect("nats://127.0.0.1:4222")
		.await
		.unwrap();	
	let subscription = connection.subscribe("test").await.unwrap();
	connection.publish("test", "no echo eh?").await.unwrap();
	assert_eq!(
		subscription.next().await.unwrap().data, 
		b"no echo eh?".to_vec()
	);
}
```

#### Expected result:

With a nats bus running at `127.0.0.1:4222` this code should hang forever as with `no_echo` set no messages should be received.

#### Actual result:

This code does not crash demonstrating a message with the same payload as the outgoing message is received despite the presence of `no_echo`.

**Comment by Jess3Jane** [comment]
If relevant this is with server version `v2.1.9`

**Comment by Jess3Jane** [comment]
I stumbled across this upon further investigation I have discovered that `publish_with_reply_or_headers` when provided with any headers, even blank headers, has the expected behavior. For example, this snippet hangs forever awaiting an incoming message as expected:

```rust
use async_nats::Options;

#[tokio::main]
async fn main() {
	let connection = Options::new()
		.no_echo()
		.connect("nats://127.0.0.1:4222")
		.await
		.unwrap();	
	let subscription = connection.subscribe("test").await.unwrap();
	connection
		.publish_with_reply_or_headers(
			"test", 
			None,
			Some(&Default::default()),
			"no echo eh?"
		)
		.await
		.unwrap();
	assert_eq!(
		subscription.next().await.unwrap().data, 
		b"no echo eh?".to_vec()
	);
}
```

**Comment by Jess3Jane** [comment]
Hm, this actually appears to be failing to echo as it causes an error on the server:

```
[452405] 2021/02/04 03:02:49.894690 [TRC] 127.0.0.1:37792 - cid:5 - <<- [CONNECT {"verbose":false,"pedantic":false,"lang":"rust","version":"0.9.2","tls_required":false,"headers":true}]
[452405] 2021/02/04 03:02:49.894766 [TRC] 127.0.0.1:37792 - cid:5 - <<- [PING]
[452405] 2021/02/04 03:02:49.894787 [TRC] 127.0.0.1:37792 - cid:5 - ->> [PONG]
[452405] 2021/02/04 03:02:49.895795 [TRC] 127.0.0.1:37792 - cid:5 - <<- [PING]
[452405] 2021/02/04 03:02:49.895824 [TRC] 127.0.0.1:37792 - cid:5 - ->> [PONG]
[452405] 2021/02/04 03:02:49.896631 [TRC] 127.0.0.1:37792 - cid:5 - <<- [SUB test 1]
[452405] 2021/02/04 03:02:49.896697 [TRC] 127.0.0.1:37792 - cid:5 - ->> [-ERR Unknown Protocol Operation]
[452405] 2021/02/04 03:02:49.896744 [ERR] 127.0.0.1:37792 - cid:5 - Client parser ERROR, state=0, i=12: proto='"HPUB test 12 23\r\nNATS/1.0\r\n\r\nno "...'
```

**Comment by spacejam** [comment]
fix is published as 0.9.4 after @stjepang 's fix was merged
